add apisix as a provider to readme and update provider guide (#123)
kubernetes-sigs/ingress2gateway: PROVIDER.md

* Familiarity with Go programming language.
* Basic understanding of Kubernetes and its custom resources.
* A setup Go development environment.

## Step-by-Step Implementation

In this section, we will walk through a demo of how to add support for the `example-gateway` provider.

1. Add a new package under the `providers` package. Say we want to add a new gateway provider example.

   ```text
   .
   ├── ingress2gateway.go
   ├── ingress2gateway_test.go
   ├── provider.go
   └── providers
       ├── common
       ├── examplegateway
       └── ingressnginx
   ```

2. Create a struct named `resourceReader` which implements the `CustomResourceReader` interface in a file named
   `resource_reader.go`.

   ```go
   package examplegateway

   import (
       "context"

       "github.com/kubernetes-sigs/ingress2gateway/pkg/i2gw/providers/common"
   )

   // converter implements the i2gw.CustomResourceReader interface.
   type resourceReader struct {
       conf *i2gw.ProviderConf
   }

   // newResourceReader returns a resourceReader instance.
   func newResourceReader(conf *i2gw.ProviderConf) *resourceReader {
       return &resourceReader{
           conf: conf,
       }
   }

>    func (r *resourceReader) readResourcesFromCluster(ctx context.Context) (*storage, error) {

Author: I used this as a reference: https://github.com/kubernetes-sigs/ingress2gateway/blob/f5324d7af13a6625bb541832f78117ba8abe2e2b/pkg/i2gw/providers/kong/resource_reader.go#L38

>    func (r *resourceReader) readResourcesFromCluster(ctx context.Context) (*storage, error) {

Reviewer: This is an internal implementation of https://github.com/kubernetes-sigs/ingress2gateway/blob/f5324d7af13a6625bb541832f78117ba8abe2e2b/pkg/i2gw/providers/kong/kong.go#L57

The provider need to implement the interfaces. Then Kong and other providers decided to have a private method to do some of the logic but `ReadResourceFromCluster` is the one that has to be implemented. In this case it just calls `readResourcesFromCluster`

>    func (r *resourceReader) readResourcesFromCluster(ctx context.Context) (*storage, error) {

Author: Ok. Thanks for clarifying.

       // read example-gateway related resources from the cluster.
       return nil, nil
   }

   func (r *resourceReader) readResourcesFromFile(ctx context.Context, filename string) (*storage, error) {
       // read example-gateway related resources from the file.
       return nil, nil
   }
   ```

>    These methods are used by providers to read and store additional resources they may need during conversion.

Reviewer: ```suggestion
   These methods are used by providers to read and store Ingresses and additional resources they may need during conversion.
```

> 3. Create a struct named `storage` in a file named `storage.go`:

Reviewer: maybe add instructions on what should be stored in storage? i.e  ingresses, CRDs or essentially more things that the provider need for the conversion

and related resources (e.g istio VirtualService) into `Gateway API` resources.
A provider must be able to read its custom resources, and convert them.

## Prerequisites

* Familiarity with Go programming language.
* Basic understanding of Kubernetes and its custom resources.
* A setup Go development environment.

## Step-by-Step Implementation

In this section, we will walk through a demo of how to add support for the `example-gateway` provider.

1. Add a new package under the `providers` package. Say we want to add a new gateway provider example.

   ```text
   .
   ├── ingress2gateway.go
   ├── ingress2gateway_test.go
   ├── provider.go
   └── providers
       ├── common
       ├── examplegateway
       └── ingressnginx
   ```

2. Create a struct named `resourceReader` which implements the `CustomResourceReader` interface in a file named
   `resource_reader.go`.

   ```go
   package examplegateway

   import (
       "context"

       "github.com/kubernetes-sigs/ingress2gateway/pkg/i2gw/providers/common"
   )

   // converter implements the i2gw.CustomResourceReader interface.
   type resourceReader struct {

>        conf *i2gw.ProviderConf

Reviewer: @pottekkat I am not sure I understand what was the trigger for changing the docs ? 
There are some inaccurate bits like this, the package in the import has changed to `.../providers/common` (why?) but this `i2gw.ProviderConf` stayed.

I know you opened this PR a while ago (apologies it goes so long). Can you please go over the whole docs change and double check everything is aligned and make sense?

Author: @LiorLieberman The reason for changing the docs was the provider code was refactored. It might be outdated now and I have lost a lot of context. I have updated the PR to just add APISIX to the provider guide. I will get to updating the docs in a separate PR after going through the code again.

	}
}

func (r *resourceReader) readResourcesFromCluster(ctx context.Context) (*storage, error) {
	// read example-gateway related resources from the cluster.

> 	return nil

Reviewer 2: ```suggestion
	return nil, nil
```

}

func (r *resourceReader) readResourcesFromFile(ctx context.Context, filename string) (*storage, error) {
	// read example-gateway related resources from the file.

> 	return nil

Reviewer 2: ```suggestion
	return nil, nil
```

	}
}
```

> 5. Create a new struct named after the provider you are implementing. This struct should embed the previous 2 structs you created.

Reviewer 2: ```suggestion
5. Create a new struct named after the provider you are implementing. This struct should embed the previous 3 structs you created.
```

```

## Creating a feature parser

In case you want to add support for the conversion of a specific feature within a provider (see for example the canary feature of ingress-nginx) you'll want to implement a `FeatureParser` function.

Different `FeatureParsers` within the same provider will run in undetermined order. This means that when building a `Gateway API` resource manifest, you cannot assume anything about previously initialized fields. The function must modify / create only the required fields of the resource manifest and nothing else.

> For example, lets say we are implementing the canary feature of some provider. When building the `HTTPRoute`, we cannot assume that the `BackendRefs` is already initialized with every `BackendRef` required. The canary `FeatureParser` function must add every missing `BackendRef` and update existing ones.

Reviewer 3: ```suggestion
For example, let's say we are implementing the canary feature of a provider. When building the `HTTPRoute`, we cannot assume that the `BackendRefs` is already initialized with every `BackendRef` required. The canary `FeatureParser` function must add every missing `BackendRef` and update existing ones.
```
